File: swift/libipc/Sources/LibIPC/Proto/Codec.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors

/// Wire-level codec identifiers for typed protocol wrappers.
public enum CodecId: UInt8, Sendable {
    case flatBuffers = 1
    case protobuf = 2
    case capnp = 3
}

/// Codec contract for typed protocol wrappers.
public protocol TypedCodec {
    associatedtype Root
    associatedtype MessageType
    associatedtype BuilderType

    static var codecId: CodecId { get }

    static func encode(builder: BuilderType) -> [UInt8]
    static func decode(buffer: IpcBuffer) -> MessageType
    static func verify(message: MessageType) -> Bool
}

public extension TypedCodec {
    static func verify(message _: MessageType) -> Bool { true }
}

File: rust/libipc/src/proto/codec.rs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya contributors
//
// Generic codec abstraction for typed protocol wrappers.

use crate::buffer::IpcBuffer;

/// Wire-level codec identifiers used by typed protocol wrappers.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum CodecId {
    FlatBuffers = 1,
    Protobuf = 2,
    Capnp = 3,
}

/// Codec contract for typed protocol wrappers.
///
/// `T` is the schema root type used by the wrapper.
pub trait Codec<T> {
    type Message;
    type Builder;

    const CODEC_ID: CodecId;

    /// Encode a pre-built message into raw bytes for transport.
    fn encode(builder: &Self::Builder) -> &[u8];

    /// Decode transport bytes into a typed message view/value.
    fn decode(buf: IpcBuffer) -> Self::Message;

    /// Optional validation hook for codec-specific checks.
    fn verify(_message: &Self::Message) -> bool {
        true
    }
}
